### DISCUSSION TITLE
How to get the event_loop from TestClient

### DISCUSSION THREAD

**Discussion by OsmelSynData** [fastapi/fastapi#10086] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
"""Example main"""
import asyncio
from fastapi import FastAPI
from api import utils

app = FastAPI()

@app.post("/save_info")
async def save_info(...):
   """Code goes here"""
   asyncio.create_task(utils.save_info)
   return {"id": "some-id"}

"""Example Tests"""
import pytest
import asyncio
from fastapi.testclient import TestClient

@pytest.fixture(scope="module")
def client(...):
  with TestClient(app) as client:
     yield client

def test_assert_save_info_tasks_finished(client):
    res = client.post("/save_info")
    assert res.status_code == 201
    """How can I assert the tasks finished from the event_loop"""
```

### Description

Hi,

Description:
I have endpoints that create `asyncio.Tasks` (background tasks). These tasks are prone to errors, and I’m also creating monitoring tasks for the background tasks. I want to test that all background tasks are finished within different scenarios (ie that we don’t have tasks that are “hanging”). How can I test these? I was thinking of getting the event loop from the `TestClient` to confirm no coroutines are pending there, is there a way to do this?

All help would be much appreciated, thanks! 

### Operating System

macOS

### FastAPI Version

0.100.1

### Pydantic Version

2.1.1

### Python Version

Python 3.11.4

**Comment by n8sty** [comment]
You can use the `event_loop` fixture from `pytest-asyncio` (https://pytest-asyncio.readthedocs.io/en/latest/reference/fixtures.html#event-loop). Or, you can create your own `event_loop` fixture, or just grab the existing or create a new event loop in your test function by using `asyncio.get_event_loop`. To create an `event_loop` fixture do something similar to the following:

```
import asyncio

@pytest.fixture
def event_loop():
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()
```
(very similar to what `pytest-asyncio` does).

I suggest using `pytest-asyncio` assuming you're using `pytest` because there's some other test collection and setup functionality it provides in addition to just the fixture above.

**Reply by OsmelSynData** [reply]
Hi thanks for the reply, I have already tried this approach and to no avail. The issue with your solution was that doing `asyncio.all_tasks(loop=event_loop)` returned the test_suite as a pending task instead of the actual background task created in the endpoint I am testing.
